### ISSUE TITLE
Docker engine 1.9.0-rc1 does not gracefully restart

### ISSUE THREAD

**Issue by BrianBland** [docker/docker#17083]
Description of problem: When stopping the docker daemon via `service docker stop`, it's possible to get in a state where the engine immediately crashes on start with a network configuration issue. This is not resolved by restarting the docker host.

`docker version`:
```
Client:
 Version:      1.9.0-rc1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   9291a0e
 Built:        Wed Oct 14 03:04:22 UTC 2015
 OS/Arch:      linux/amd64
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```
(Server version is same as client)

`docker info`:
?

`uname -a`:
`Linux vagrant-ubuntu-trusty-64 3.13.0-39-generic #66-Ubuntu SMP Tue Oct 28 13:30:27 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux`

Environment details (AWS, VirtualBox, physical, etc.):
Ubuntu 14.04, VirtualBox (vagrant)

How reproducible:
Not sure, happened on the 2nd time I restarted my daemon after upgrading from 1.8.2 to 1.9.0-rc1

Steps to Reproduce:
1. Run a container
2. Stop docker daemon
3. Try starting docker daemon

Actual Results:
```bash
$ sudo service docker start
docker start/running, process 19174
$ sudo docker ps
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
$ sudo tail /var/log/upstart/docker.log
/var/run/docker.sock is up
INFO[0000] [graphdriver] using prior storage driver "aufs" 
INFO[0000] API listen on /var/run/docker.sock           
INFO[0000] Firewalld running: false                     
FATA[0000] Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints 
```

Expected Results:
Docker daemon does not crash immediately

Additional info:

**Comment by abronan** [comment]
ping @mavenugo @mrjana 

@icecrime, seems like the issue you encountered.

**Comment by mrjana** [comment]
thanks @BrianBland  for the report.

It looks like you were running linked containers and in case when restarting the docker daemon one of the linked container failed to load from hard disk so when the parent container was getting restarted the network plumbing went as far as creating an endpoint but failed to find the child container so it bailed out but the code should have deleted the endpoint this is what is causing the docker daemon to not restart next time. Will push a PR to fix this.

**Comment by LK4D4** [comment]
I had same issue. Howerver I tried to stop daemon with ~300 running containers.

**Comment by guss77** [comment]
I have the same problem, with a more recent docker installation:

~~~
# docker version
Client:
 Version:      1.9.0-dev
 API version:  1.22
 Go version:   go1.4.3
 Git commit:   8d62f24
 Built:        Thu Oct 29 20:16:32 UTC 2015
 OS/Arch:      linux/amd64
 Experimental: true
Error response from daemon: 404 page not found
~~~

And the docker daemon fails to start:

~~~
$ /usr/bin/docker daemon -H unix:///var/run/docker.sock
WARN[0000] Running experimental build
INFO[0000] API listen on /var/run/docker.sock
INFO[0000] [graphdriver] using prior storage driver "aufs"
INFO[0000] Firewalld running: false
WARN[0000] Could not delete local endpoint compassionate_bassi during endpoint cleanup: endpoint with name compassionate_bassi id 4bcca52afd583539110a25fe98034eec680f01d2e1762e5bf558e8cb13129c8c has active containers
FATA[0000] Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints
~~~

Running docker-engine experimental on Ubuntu 15.10. The issue happened after I stopped a build that was in progress. after that the server crashed and would not restart. Rebooting the host did not solve the problem.

**Comment by guss77** [comment]
Moving /var/lib/docker away solved the problem. I have the original storage if anyone wants to look at it.

**Comment by ruFog** [comment]
I'd updated docker from 1.8 to 1.9 (Ubuntu 14.04) and had the same problem. Thanks to @guss77 i removed only /var/lib/docker/network and now everything works well and without containers lost.

**Comment by githubwjj** [comment]
@ruFog   thank you!

**Comment by fredzica** [comment]
Removing the '/var/lib/docker/network' folder worked for me.

**Comment by Skyross** [comment]
@fredzica Worked same for me here.

**Comment by fxposter** [comment]
The issue is still there in 1.9.0 release.
Debian 8. Kernel 3.16.

**Comment by schmunk42** [comment]
I am seeing this with AWS machines, swarm and consul backend.

Node config:

```
ubuntu@sepp:~$ uname -a
Linux sepp 3.13.0-53-generic #89-Ubuntu SMP Wed May 20 10:34:39 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
ubuntu@sepp:~$ docker --version
Docker version 1.9.0, build 76d6bc9
```

Trying to remove a network:

```
[oot@sepp-roj:~# docker network inspect playground3
    {
        "Name": "playground3",
        "Id": "f182d2dbcc723764f9b8d71b175315ed05d687a444baa95836139eed93243711",
        "Scope": "global",
        "Driver": "overlay",
        "IPAM": {
            "Driver": "default",
            "Config": [
                {}
            ]
        },
        "Containers": {},
        "Options": {}
    }
]
root@sepp-roj:~# docker network rm playground3
Error response from daemon: 500 Internal Server Error: network playground3 has active endpoints
```

**Comment by mavenugo** [comment]
@schmunk42 couple of related points -
1. overlay network driver requires kernel version 3.16+
2. this issue is resolved in 1.9.1. RC1 is out.

Can you pls upgrade the kernel and try the 1.9.1-RC1 binary ?

**Comment by schmunk42** [comment]
@mavenugo The network removal is indeed fixed in 1.9.1-rc1, thank you.
Still working on getting a newer kernel on AWS...

**Comment by tiborvass** [comment]
@schmunk42 thanks for the confirmation!

**Comment by defra91** [comment]
@ruFog Removing `/var/lib/docker/network` also worked for me. Additional details:

* Ubuntu 15.04 Vivid
* Docker version 1.9.0, build 76d6bc9

**Comment by schmunk42** [comment]
just FYI: My problems above were solved with 1.9.1-rc1.

**Comment by luca3m** [comment]
I have this problem on Ubuntu 15.10 and docker 1.9.0 build `76d6bc9`, same of @defra91

**Comment by danehammer** [comment]
>> 1. overlay network driver requires kernel version 3.16+

Is there a way this could get called out more strongly? I have a swarm with different OSes and this seems to explain me seeing the issue. But why was I able to point my docker client at the swarm master process, and then create an overlay network at all? Maybe it could fail on the node(s) with a kernel version too low, but on a node with a kernel of 3.10.80-121, I would not expect to see `docker network ls` return a network with an overlay driver. You would think it would refuse to create it if the dependencies aren't met.

**Comment by thaJeztah** [comment]
@danehammer can you open a new issue for that? If you have ideas on where you think that information should be added, mention that in the description of your issue (or perhaps you're interested in creating a pull request, the markdown files used to create the documentation can be found in this repository; https://github.com/docker/docker/tree/master/docs :+1:)

**Comment by ddproxy** [comment]
I'm also having this issue with Docker 1.9.1

````
docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Tue 2015-11-24 15:10:30 MST; 3s ago
     Docs: https://docs.docker.com
  Process: 6512 ExecStart=/usr/bin/docker daemon -H fd:// (code=exited, status=1/FAILURE)
 Main PID: 6512 (code=exited, status=1/FAILURE)

Nov 24 15:10:30 jon-M11AD systemd[1]: Starting Docker Application Container Engine...
Nov 24 15:10:30 jon-M11AD docker[6512]: time="2015-11-24T15:10:30.202538083-07:00" level=info msg="API listen on /var/run/docker.sock"
Nov 24 15:10:30 jon-M11AD docker[6512]: time="2015-11-24T15:10:30.204322667-07:00" level=info msg="[graphdriver] using prior storage driver \"aufs\""
Nov 24 15:10:30 jon-M11AD docker[6512]: time="2015-11-24T15:10:30.267340307-07:00" level=info msg="Firewalld running: false"
Nov 24 15:10:30 jon-M11AD docker[6512]: time="2015-11-24T15:10:30.291595304-07:00" level=fatal msg="Error starting daemon: Error initializing network controller: could not delete the default bridge network: network ...active endpoints"
Nov 24 15:10:30 jon-M11AD systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Nov 24 15:10:30 jon-M11AD systemd[1]: Failed to start Docker Application Container Engine.
Nov 24 15:10:30 jon-M11AD systemd[1]: docker.service: Unit entered failed state.
Nov 24 15:10:30 jon-M11AD systemd[1]: docker.service: Failed with result 'exit-code'.
Hint: Some lines were ellipsized, use -l to show in full.

$ docker -v
Docker version 1.9.1, build a34a1d5

````

Removing /var/lib/docker/network allowed me to restart the daemon.

**Additional Information**
I just performed a restart of my machine to finish updates (Ubuntu 15.10) and did not gracefully stop my docker containers - which wasn't an issue before 1.9.1 was installed.

**Comment by thaJeztah** [comment]
@ddproxy when you restarted your machine, was it still running 1.9.0? this looks related to https://github.com/docker/docker/issues/17555 (fixed by https://github.com/docker/docker/pull/17558)

**Comment by ddproxy** [comment]
````
Start-Date: 2015-11-23  10:40:55
Commandline: aptdaemon role='role-commit-packages' sender=':1.394'
Upgrade: kpartx:amd64 (0.5.0-7ubuntu7, 0.5.0-7ubuntu7.1), im-config:amd64 (0.29-1ubuntu7, 0.29-1ubuntu7.1), libpng3:amd64 (1.2.51-0ubuntu3, 1.2.51-0ubuntu3.15.10.1), whoopsie:amd64 (0.2.49, 0.2.49.1), libpng12-0:amd64 (1.2.51-0ubuntu3, 1.2.51-0ubuntu3.15.10.1), libtracker-sparql-1.0-0:amd64 (1.4.1-1ubuntu2, 1.4.1-1ubuntu2.1), libwhoopsie0:amd64 (0.2.49, 0.2.49.1), libpng12-dev:amd64 (1.2.51-0ubuntu3, 1.2.51-0ubuntu3.15.10.1), lightdm:amd64 (1.16.4-0ubuntu1, 1.16.5-0ubuntu1), liblightdm-gobject-1-0:amd64 (1.16.4-0ubuntu1, 1.16.5-0ubuntu1), docker-engine:amd64 (1.9.0-0~wily, 1.9.1-0~wily)
End-Date: 2015-11-23  10:41:31
````

Yup, would have been running 1.9.0 and restarted as 1.9.1

**Comment by caseca** [comment]
Thenk you!

**Comment by schmunk42** [comment]
Same here with 1.9.1, removing networks, like @ddproxy mentioned helps, see also http://stackoverflow.com/a/33604859/291573

```
docker@runner-host-5:~$ docker info
Containers: 34
Images: 979
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1047
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.13-boot2docker
Operating System: Boot2Docker 1.9.1 (TCL 6.4.1); master : cef800b - Fri Nov 20 19:33:59 UTC 2015
CPUs: 4
Total Memory: 1.463 GiB
Name: runner-host-5
ID: NGM6:QJOX:QWRO:MCIG:N45R:PPNV:KYPX:NIIM:45BT:LVUK:TNK4:HUO5
Debug mode (server): true
 File Descriptors: 28
 Goroutines: 56
 System Time: 2015-12-09T03:51:11.16999533Z
 EventsListeners: 0
 Init SHA1: 
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

**Comment by waynew** [comment]
Running Docker on Arch linux with the same problem, and removing `/var/lib/docker/networks` also worked:

```
$ docker --version
Docker version 1.9.1, build a34a1d5-dirty
$ uname -a
Linux myhost 4.2.5-1-ARCH #1 SMP PREEMPT Tue Oct 27 08:13:28 CET 2015 x86_64 GNU/Linux
```

**Comment by cbastienbaron** [comment]
same here `mv /var/lib/docker/network/files  /var/lib/docker/network/files_bck`  :+1: 

seems to be work for error mentionned by daemon : 

`FATA[0000] Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints `

```
uname -a
Linux bastien-cws 3.13.0-74-generic #118-Ubuntu SMP Thu Dec 17 22:52:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

docker --version
Docker version 1.9.1, build a34a1d5
```

**Comment by ghasemz** [comment]
sudo rm -rf  /var/lib/docker/network/files

solved!

**Comment by OmPrakash22** [comment]
Removing the '/var/lib/docker' folder worked for me.

**Comment by dieend** [comment]
Removing `/var/lib/docker/network` works for me. Thanks!

**Comment by Rolyat971** [comment]
Thanks @ghasemz :)

**Comment by mgcrea** [comment]
Just encountered this after an update to `1.10`. This should probably be reopened.

**Comment by at88mph** [comment]
Ditto for 1.10.  Removing /var/lib/docker/network worked for me on Debian Stretch.

**Comment by pedrommone** [comment]
Removing `var/lib/docker/network` also worked for me.

```
root@10:/var/lib/docker# uname -a
Linux 10 3.13.0-74-generic #118-Ubuntu SMP Thu Dec 17 22:52:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
root@10:/var/lib/docker# docker --version
Docker version 1.9.1, build a34a1d5
```

**Comment by jonvargas** [comment]
Removing network folder worked for me:

Kernel: 4.2.0-27-generic
Distro: Ubuntu 15.10
Docker: 1.10.1, build 9e83765

**Comment by thaJeztah** [comment]
@jonvargas was that on a fresh 1.10.1 install, or did you upgrade from an earlier version of Docker? Improvements should be made in docker 1.10 to prevent these issues, but if you upgraded from an earlier version of docker, this may have been left-over issues from the old version.

**Comment by Titotix** [comment]
I am on a fresh 1.10.1 and I got the same issue which is fix by removing `/var/lib/docker/network`.
```
uname -a
Linux demoniac 4.3.0-0.bpo.1-amd64 #1 SMP Debian 4.3.3-7~bpo8+1 (2016-01-19) x86_64 GNU/Linux
```

**Comment by jonvargas** [comment]
I performed a dist-upgrade, maybe it affected it, because it was working
last days. But I can't confirm now.

Jonathan Vargas
Software Entrepreneur
jonathan.vargas.cr

**Comment by cxxly** [comment]
Upgrade from 1.9 to 1.10.1 and got same issue which is fix by removing /var/lib/docker/network.

**Comment by iam-peekay** [comment]
@thaJeztah I have this error with a fresh 1.10.2 install. 

I tried all the mentioned solutions in this thread and other threads:
1) Remove the ```/var/lib/docker/network``` folder
2) To create a docker group with a user, then log out and log in and restart docker deamon
3) To uninstall and reinstall Docker

and neither of these worked for me.

**Comment by Hokutosei** [comment]
sudo rm -rf /var/lib/docker/network/files

worked to me too! thanks!!! whew I thought I was going to lost my db's!

**Comment by elicoidal** [comment]
@iam-peekay I was completely stuck (multiple unsuccessful install/uninstall etc.) and we found a strange way to solve the issue:
1. completely uninstall docker (1.10.2 in our case on Ubuntu LTS 14.04)
2. reboot the server
3. install docker 1.10.2
4. Solved our issue

**Comment by thaJeztah** [comment]
@elicoidal was this after an upgrade from 1.9 or a fresh install?

**Comment by elicoidal** [comment]
@thaJeztah docker couldnot stop anymore after a physical restart of our server (it was looping crying for docker0 nic).
Then we spent countable hours upgrading ubuntu to 14.04, installing/reinstalling 1.9, 1.10 without success.
It worked again after uninstalling docker/shutting down server/reinstall 1.10.2.
@iam-peekay did you physically restart your server?
Thanks for the trick: we will probably apply as we need to be able to restart servers at will and having docker properly shutdown and restart smoothly when physical restart.

**Comment by iam-peekay** [comment]
@elicoidal yes, I did in fact do a complete reboot of the server several times but didn't have much luck.

Glad to hear the trick might be helpful! :)

**Comment by jsdevel** [comment]
I'm affected by this right now.  I think I'm leaning toward trying what @iam-peekay did.  Removing `/var/lib/docker` isn't working for me.

**Comment by jsdevel** [comment]
Are there plans to release `1.10.0` for fedora soon?

**Comment by thaJeztah** [comment]
@jsdevel 1.10.2 for fedora is available from our repositories, be aware that Fedora 21 is EOL, and no longer supported, so if you're on Fedora 21, you need to upgrade https://docs.docker.com/engine/installation/linux/fedora/

**Comment by brettdh** [comment]
Another vote for re-opening this issue, as obviously it's occurring for lots of people, including me (generic driver). Removing `/var/lib/docker/network/files` and restarting the docker service on all swarm nodes works for me, but this is a real impediment and occurs fairly commonly when I need to remove and re-create containers with docker-compose.

**Comment by jsdevel** [comment]
Just to add some more details:

* The file that needed to be deleted for me was `local-kv.db`.
* If you've started `docker daemon` with the `-g` flag, you need to delete `./network/files/local-kv.db` in that location.
* Removing this file _will_ wipe away any networks you've created via `docker network create`.  Fortunately I only had one that was easy to re-create.

I'm glad to be up and running with docker again!

**Comment by adolphlwq** [comment]
I run into this problem too, below is my environment:
- docker info
```shell
Containers: 13
 Running: 0
 Paused: 0
 Stopped: 13
Images: 29
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 126
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.19.0-25-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.686 GiB
Name: dev
ID: FCBP:VLS2:VF74:FFHW:E2MG:W6KX:VHAZ:5Z3J:BNCQ:4N7Y:BJZ7:WL6H
Username: adolphlwq
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```
- docker version
```shell
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64
```
- uname -a
```shell
Linux dev 3.19.0-25-generic #26~14.04.1-Ubuntu SMP Fri Jul 24 21:16:20 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

And I solve this problem by removing `/var/lib/docker/network` too.
Thanks @ruFog

**Comment by svetlin-mladenov** [comment]
I am affected by the same issue. `Docker version 1.10.2, build c3959b1`.

**Comment by onorua** [comment]
The same here:
```
Mar 17 11:54:48 ip-172-16-236-27 docker[6713]: time="2016-03-17T11:54:48.089138508Z" level=fatal msg="Error starting daemon: Error initializing network controller: could not delete the default bridge network: network bridge has active endpoints"
```

removing of the `/var/lib/docker/network` allowed me to start it up, but then it just crash again..

**Comment by onorua** [comment]
@jfrazelle @abronan would you mind to reopen the bug?

**Comment by mavenugo** [comment]
@onorua the issue that you are seeing (as per your logs) is unrelated to this.

@adolphlwq @svetlin-mladenov @jsdevel @brettdh @iam-peekay and others who are seeing this issue, we have resolved a few related issues recently via #21261 and that solves a couple of cases that fixes the `active endpoint` issue during ungraceful daemon shutdown scenario. Can you please try master and give us your feedback ? And if you still see the problem, pls share the file `local-kv.db`, `docker version`, `docker info` and the daemon logs.

**Comment by brettdh** [comment]
@mavenugo is there a reliable way to reproduce this? It's been pretty sporadic for me (weeks to months), so I won't really know whether the fixes in #21261 actually solved the problem.

I could switch to master if it occurs again, but I'm not sure that would tell you anything, since it sounds like once the ungraceful shutdown has occurred, you're already in the bad state and it's too late to switch to master.

**Comment by mavenugo** [comment]
@brettdh the fixes that we have added solves couple of sporadic cases. The sporadic cases happens during ungraceful daemon shutdown which results in some inconsistent states.
